File: TcnnNerfApi/optimizer.cpp
#include "optimizer.h"

namespace tcnnNerf {
	Optimizer::Optimizer(const nlohmann::json& config) {
		m_optimizer.reset(tcnn::create_optimizer<ngp::nerf_precision>(config));
	}
	void Optimizer::allocate(uint32_t n_params, std::vector<std::pair<uint32_t, uint32_t>> layer_sizes) {
		m_optimizer->allocate(n_params, layer_sizes);
	}
	void Optimizer::step(float loss_scale, torch::Tensor weights, torch::Tensor gradients) {

		m_optimizer->step(nullptr, loss_scale, nullptr, (ngp::nerf_precision*)void_data_ptr(weights), (ngp::nerf_precision*)void_data_ptr(gradients));
	}
	nlohmann::json Optimizer::hyperparams() {
		return m_optimizer->hyperparams();
	}
}

File: InstantNeRF/Optimizer.cs
using System.IO;
using System.Runtime.InteropServices;
using TorchSharp;
using static TorchSharp.torch;
using TorchSharp.Modules;

namespace InstantNeRF
{
    public class Optimizer
    {
        private IntPtr handle;

        private Parameter param;
        private Parameter paramFP;
        public Optimizer(string config, MLP mlp)
        {
            handle = TcnnWrapper.createOptimizer(config);

            // Store a copy of full precision parameters to accurately perform parameter updates
            (param, paramFP) = mlp.getParameters();

            //allocate memory by retrieving the network layout directly in CUDA
            allocate(mlp.getHandle());
        }

        public void step(Tensor gradients)
        {
            float lossScale = param.dtype == float16 ? 128.0f : 1.0f;
            TcnnWrapper.step(handle, lossScale, param.Handle, paramFP.Handle, gradients.Handle);
        }

        public string hyperparams()
        {
            return TcnnWrapper.hyperparams(handle);
        }
        private void allocate(IntPtr module)
        {
            TcnnWrapper.allocate(handle, module);
        }

        public Parameter getParameter()
        {
            return param;
        }

    }
}